### DISCUSSION TITLE
Connection management

### DISCUSSION THREAD

**Discussion by cisaacson** [rusqlite/rusqlite#1226] [Questions]
We would like to keep a connection open for the life of the application. We are currently using `r2d2_sqlite` as a connection pool, and that works OK but we really do not need more than one connection. We tried `lazy_static` with a `Mutex` and it is not compatible with the other fn calls we need to support.

Is there a recommended way to do this with `rusqlite`?

**Comment by gwenn** [comment]
Could you please be more specific on: `it is not compatible with the other fn calls we need to support.` ?
Do you have some `rustc` errors / diagnostics ?

**Comment by cisaacson** [comment]
@gwenn No diagnostics, it was just passing references around and problems with the borrow checker.

We did make it work (see below), but I would like to know if you think a connection pool with `r2d2` is best, or just using a single connection where required behind a `Mutex`?

FYI what started this is because `r2d2` has a default timeout that closes connections, SQLite was cleaning up and removing the WAL file when the Rust connections closed. We had another persistent connection from our application layer and it was not recognized by SQLite and it caused errors on the application side.

Now we have both the connection pool (with timeouts disabled) and the single `lazy_static` connection working. Please let me know which approach you think is best for the long run.

If useful for others, we did make it work with `lazy_static` like this:

```
    static ref SQLITE_CONN: Mutex<SqliteConnection> = {
        let sqlite_path = SQLITE_PATH.to_owned();
        let sqlite_conn = SqliteConnection::open(&sqlite_path);
        let sqlite_conn = match sqlite_conn {
            Ok(v) => v,
            Err(e) => {
                panic!("Could not create the SQLite connection. sqlite_dir_path: {:?} error: {:?}", &sqlite_path, e);
            }
        };

        Mutex::new(sqlite_conn)
    };

    INIT.call_once(|| {
        let sqlite_path = SQLITE_PATH.to_owned();

        let sqlite_conn = SQLITE_CONN
            .lock()
            .map_err(|e| format!("Could not get SQLite connection. e: {:?}", e))
            .unwrap();
...
```
Then we expose a fn like this:
```
pub fn get_sqlite_conn() -> Result<MutexGuard<'static, SqliteConnection>, Box<dyn Error>> {
    let sqlite_conn = SQLITE_CONN
        .lock()
        .map_err(|e| format!("Could not get SQLite connection. e: {:?}", e))?;

    Ok(sqlite_conn)
}

```
Anywhere we need a connection in the code we just call the fn above and it works fine.

And `rusqlite` is really great, easy to use and does what we need.

**Comment by gwenn** [comment]
> Please let me know which approach you think is best for the long run.

I am afraid there is no best solution. For example,
See https://github.com/rusqlite/rusqlite/issues/342#issuecomment-592662942
And https://github.com/rusqlite/rusqlite/issues/188#issuecomment-787390455

**Comment by cisaacson** [comment]
Thanks @gwenn. It does seem like a connection pool is better then. We will stick with that, `r2d2_sqlite` is very easy to use and works well. It was important in our case to disable the idle and max lifetime timeouts, other than that it was easy to implement and we will stick with that.

Closing the issue now.
